tutorbackup/templates/backup/build/backup/s3_client.py

```diff
@@ -23,3 +24,16 @@
 
 class IntegrityError(BaseException):
     pass
+
+
+def calculate_checksum(file_name):
+    # To avoid reading in the whole file into memory, break it
+    # down into chunks and calculate the checksum by updating
+    # the md5 hash. The number of read bytes should be a
+    # multiple of the hash algorithm's block size.
+    num_of_blocks = 128
+    with open(file_name, "rb") as f:
+        file_hash = hashlib.md5()
+        while chunk := f.read(num_of_blocks * file_hash.block_size):
```

Comment:
I must confess that I am only vaguely familiar with the walrus operator, so for the benefit of my education would you mind explaining what this line does, exactly? :)

Reply:
> I must confess that I am only vaguely familiar with the walrus operator, so for the benefit of my education would you mind explaining what this while line does, exactly? :)

That line reads a number of bytes from the file and puts it in the `chunk` variable at every iteration of the loop. While `chunk` is not `None`, we update the hash. If we reach the end of the file, `chunk` will be `None` and we break out of the `while` loop.

Reply:
> That line reads a number of bytes from the file and puts it in the `chunk` variable at every iteration of the loop. While `chunk` is not `None`, we update the hash. If we reach the end of the file, `chunk` will be `None` and we break out of the `while` loop.

Got it, thank you!

Reply:
> Also, if I read the parenthetical expression correctly, then we would be reading the file in chunks of 16KiB (since MD5's digest block size is 128). This will no doubt be very lenient on memory, but I don't know if for multi-GiB tar file like the one we saw in https://github.com/hastexo/tutor-contrib-backup/issues/40, this might make the process unreasonably slow. Maybe use a higher value for num_of_blocks? Or would this take only a matter of seconds even for a multi-GiB tarball?

Yes, reading files in chunks will slow down the process. I can try testing different chunk sizes on a large file and see the effect.

Reply:
I tried different number of chunks on a 20 GiB file on my machine. My laptop fan was pretty busy for a while there :sweat_smile: :fire:
Here are the results:

number of chunks | time | improvement
--- | --- | ---
128 | 273.262 s| -
256 | 268.562 s| 1.7 %
1280 | 262.951 s| 3.4 %

So increasing the number of chunks beyond 128 doesn't really make a big impact on the total time it takes to calculate the checksum.

Reply:
OK, fair. 10 seconds give or take really doesn't make a difference. In that case, please just add a code comment indicating that the choice of 128 is, essentially, arbitrary and that it doesn't add much tangible performance benefit to select a higher number.

Reply:
will do